{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "machine_shape": "hm",
      "gpuType": "V28",
      "mount_file_id": "1wFx74CFxO_qDFRD8MfDxqWJ3ZiaOkBXA",
      "authorship_tag": "ABX9TyOc6na9FKQ3IQAJawdwwn0R",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    },
    "accelerator": "TPU"
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/yc386/orthrus_metaproteomics/blob/main/annotated_orthrus_pt2.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# **Orthrus pt2- [`Sage`](https://github.com/lazear/sage) + [`Mokapot`](https://github.com/wfondrie/mokapot)**\n",
        "\n",
        "Please note: change to **TPU** runtime if RAM usage is expected to be high (due to files/PTMs/databases)"
      ],
      "metadata": {
        "id": "g529CpYsDdjc"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Add inputs for `SAGE` -> click `Runtime` -> `Run all`\n",
        "\n",
        "#@markdown **Parameters for `SAGE`**\n",
        "peak_folder = \"/content/drive/Shareddrives/MS_Palaeome/Odense_RAWs/23_Apr_2024/mzML/test\" #@param {type:\"string\"}\n",
        "file_type=\"mzML\" #@param [\"mzML\", \"mgf\"]\n",
        "#@markdown - use the drop-down menu to choose the instrument file type\n",
        "\n",
        "#@markdown **Option 1: `SAGE` PTM plus**\n",
        "#@markdown - Default `Sage` contains\n",
        "use_PTM_plus = True #@param {type:\"boolean\"}\n",
        "max_variable_mods= 3 #@param {type:\"number\"}\n",
        "missed_cleavages= 2 #@param {type:\"number\"}\n",
        "AA_1 = \"M\" #@param [\"None\", \"[\",\"]\",\"A\", \"C\", \"D\", \"E\", \"F\", \"G\", \"H\", \"I\", \"K\", \"L\", \"M\", \"N\", \"P\", \"Q\", \"R\", \"S\", \"T\", \"V\", \"W\", \"Y\"]\n",
        "AA_1_mod = 15.9949 #@param {type:\"number\"}\n",
        "AA_2 = \"P\" #@param [\"None\", \"[\",\"]\",\"A\", \"C\", \"D\", \"E\", \"F\", \"G\", \"H\", \"I\", \"K\", \"L\", \"M\", \"N\", \"P\", \"Q\", \"R\", \"S\", \"T\", \"V\", \"W\", \"Y\"]\n",
        "AA_2_mod = 15.9949 #@param {type:\"number\"}\n",
        "AA_3 = \"N\" #@param [\"None\", \"[\",\"]\",\"A\", \"C\", \"D\", \"E\", \"F\", \"G\", \"H\", \"I\", \"K\", \"L\", \"M\", \"N\", \"P\", \"Q\", \"R\", \"S\", \"T\", \"V\", \"W\", \"Y\"]\n",
        "AA_3_mod = 0.984016 #@param {type:\"number\"}\n",
        "AA_4 = \"Q\" #@param [\"None\", \"[\",\"]\",\"A\", \"C\", \"D\", \"E\", \"F\", \"G\", \"H\", \"I\", \"K\", \"L\", \"M\", \"N\", \"P\", \"Q\", \"R\", \"S\", \"T\", \"V\", \"W\", \"Y\"]\n",
        "AA_4_mod = 0.984016 #@param {type:\"number\"}\n",
        "AA_5 = \"None\" #@param [\"None\", \"[\",\"]\",\"A\", \"C\", \"D\", \"E\", \"F\", \"G\", \"H\", \"I\", \"K\", \"L\", \"M\", \"N\", \"P\", \"Q\", \"R\", \"S\", \"T\", \"V\", \"W\", \"Y\"]\n",
        "#@markdown - [ = n-terminal\n",
        "AA_5_mod = \t42.010565 #@param {type:\"number\"}\n",
        "\n",
        "#@markdown **Option 2: `Mokapot` rescoring**\n",
        "use_Mokapot = True #@param {type:\"boolean\"}\n",
        "#@markdown - machine learning-based rescoring per experiment or across experiments\n",
        "joint_modelling= False #@param {type:\"boolean\"}\n",
        "#@markdown - a joint model for low abundance samples\n",
        "default_Percolator=False #@param {type:\"boolean\"}\n",
        "#@markdown - Python implementation of the Percolator SVM model"
      ],
      "metadata": {
        "id": "QSA5zC6q6knU",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "6KE6TaQj3oxE",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "cellView": "form",
        "outputId": "a9f4c213-f21c-4d8f-f9b0-b1af41a14092"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting mokapot\n",
            "  Downloading mokapot-0.10.0-py3-none-any.whl.metadata (4.5 kB)\n",
            "Requirement already satisfied: xgboost in /usr/local/lib/python3.10/dist-packages (2.1.2)\n",
            "Requirement already satisfied: numpy>=1.18.1 in /usr/local/lib/python3.10/dist-packages (from mokapot) (1.26.4)\n",
            "Requirement already satisfied: pandas>=1.0.3 in /usr/local/lib/python3.10/dist-packages (from mokapot) (2.2.2)\n",
            "Requirement already satisfied: scikit-learn>=0.22.1 in /usr/local/lib/python3.10/dist-packages (from mokapot) (1.5.2)\n",
            "Requirement already satisfied: numba>=0.48.0 in /usr/local/lib/python3.10/dist-packages (from mokapot) (0.60.0)\n",
            "Requirement already satisfied: matplotlib>=3.1.3 in /usr/local/lib/python3.10/dist-packages (from mokapot) (3.8.0)\n",
            "Requirement already satisfied: lxml>=4.6.2 in /usr/local/lib/python3.10/dist-packages (from mokapot) (5.3.0)\n",
            "Collecting triqler>=0.6.2 (from mokapot)\n",
            "  Downloading triqler-0.7.3-py3-none-any.whl.metadata (11 kB)\n",
            "Requirement already satisfied: joblib>=1.1.0 in /usr/local/lib/python3.10/dist-packages (from mokapot) (1.4.2)\n",
            "Requirement already satisfied: importlib-metadata>=5.1.0 in /usr/local/lib/python3.10/dist-packages (from mokapot) (8.5.0)\n",
            "Requirement already satisfied: nvidia-nccl-cu12 in /usr/local/lib/python3.10/dist-packages (from xgboost) (2.23.4)\n",
            "Requirement already satisfied: scipy in /usr/local/lib/python3.10/dist-packages (from xgboost) (1.13.1)\n",
            "Requirement already satisfied: zipp>=3.20 in /usr/local/lib/python3.10/dist-packages (from importlib-metadata>=5.1.0->mokapot) (3.20.2)\n",
            "Requirement already satisfied: contourpy>=1.0.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=3.1.3->mokapot) (1.3.0)\n",
            "Requirement already satisfied: cycler>=0.10 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=3.1.3->mokapot) (0.12.1)\n",
            "Requirement already satisfied: fonttools>=4.22.0 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=3.1.3->mokapot) (4.54.1)\n",
            "Requirement already satisfied: kiwisolver>=1.0.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=3.1.3->mokapot) (1.4.7)\n",
            "Requirement already satisfied: packaging>=20.0 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=3.1.3->mokapot) (24.1)\n",
            "Requirement already satisfied: pillow>=6.2.0 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=3.1.3->mokapot) (10.4.0)\n",
            "Requirement already satisfied: pyparsing>=2.3.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=3.1.3->mokapot) (3.2.0)\n",
            "Requirement already satisfied: python-dateutil>=2.7 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=3.1.3->mokapot) (2.8.2)\n",
            "Requirement already satisfied: llvmlite<0.44,>=0.43.0dev0 in /usr/local/lib/python3.10/dist-packages (from numba>=0.48.0->mokapot) (0.43.0)\n",
            "Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.10/dist-packages (from pandas>=1.0.3->mokapot) (2024.2)\n",
            "Requirement already satisfied: tzdata>=2022.7 in /usr/local/lib/python3.10/dist-packages (from pandas>=1.0.3->mokapot) (2024.2)\n",
            "Requirement already satisfied: threadpoolctl>=3.1.0 in /usr/local/lib/python3.10/dist-packages (from scikit-learn>=0.22.1->mokapot) (3.5.0)\n",
            "Collecting job-pool<0.3.0,>=0.2.6 (from triqler>=0.6.2->mokapot)\n",
            "  Downloading job_pool-0.2.6-py3-none-any.whl.metadata (2.1 kB)\n",
            "Requirement already satisfied: tqdm<5.0.0,>=4.66.1 in /usr/local/lib/python3.10/dist-packages (from job-pool<0.3.0,>=0.2.6->triqler>=0.6.2->mokapot) (4.66.6)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.10/dist-packages (from python-dateutil>=2.7->matplotlib>=3.1.3->mokapot) (1.16.0)\n",
            "Downloading mokapot-0.10.0-py3-none-any.whl (74 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m74.4/74.4 kB\u001b[0m \u001b[31m7.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading triqler-0.7.3-py3-none-any.whl (64 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m64.2/64.2 kB\u001b[0m \u001b[31m6.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading job_pool-0.2.6-py3-none-any.whl (9.1 kB)\n",
            "Installing collected packages: job-pool, triqler, mokapot\n",
            "Successfully installed job-pool-0.2.6 mokapot-0.10.0 triqler-0.7.3\n",
            "installing conda...\n",
            "CPU times: user 420 ms, sys: 77 ms, total: 497 ms\n",
            "Wall time: 1min 30s\n"
          ]
        }
      ],
      "source": [
        "#@title install dependencies\n",
        "\n",
        "%%time\n",
        "\n",
        "import os\n",
        "\n",
        "!pip install mokapot xgboost\n",
        "\n",
        "if not os.path.isfile(\"Sage_READY\"):\n",
        "  print(\"installing conda...\")\n",
        "  os.system(\"wget -qnc https://github.com/conda-forge/miniforge/releases/latest/download/Mambaforge-Linux-x86_64.sh\")\n",
        "  os.system(\"bash Mambaforge-Linux-x86_64.sh -bfp /usr/local\")\n",
        "  os.system(\"touch Sage_READY\")\n",
        "  os.system(f\"conda install -c bioconda -c conda-forge sage-proteomics -y -q\")\n",
        "\n",
        "import shutil\n",
        "\n",
        "'''\n",
        "Organise mztabs and instrument files in the same folder\n",
        "\n",
        "'''\n",
        "\n",
        "def organise_files (directory):\n",
        "\n",
        "    if not os.path.isdir(directory):\n",
        "        print(f\"The directory {directory} does not exist.\")\n",
        "        return\n",
        "\n",
        "    mzml_files = glob.glob(os.path.join(directory, '*.mzML'))\n",
        "\n",
        "    for mzml_file in mzml_files:\n",
        "\n",
        "        base_name = os.path.splitext(os.path.basename(mzml_file))[0]\n",
        "\n",
        "        new_folder_path = os.path.join(directory, base_name)\n",
        "\n",
        "        if not os.path.exists(new_folder_path):\n",
        "            os.makedirs(new_folder_path)\n",
        "            print(f\"Created folder: {new_folder_path}\")\n",
        "        else:\n",
        "            print(f\"Folder already exists: {new_folder_path}\")\n",
        "\n",
        "\n",
        "        new_mzml_path = os.path.join(new_folder_path, os.path.basename(mzml_file))\n",
        "        if not os.path.exists(new_mzml_path):\n",
        "            shutil.move(mzml_file, new_folder_path)\n",
        "            print(f\"Moved {mzml_file} to {new_folder_path}\")\n",
        "        else:\n",
        "            print(f\".mzML file already exists in the destination: {new_mzml_path}\")\n",
        "\n",
        "\n",
        "        fasta_filename = f\"{base_name}_matched.fasta\"\n",
        "        fasta_file = os.path.join(directory, fasta_filename)\n",
        "\n",
        "\n",
        "        if os.path.exists(fasta_file):\n",
        "            new_fasta_path = os.path.join(new_folder_path, fasta_filename)\n",
        "            if not os.path.exists(new_fasta_path):\n",
        "                shutil.move(fasta_file, new_folder_path)\n",
        "                print(f\"Moved {fasta_file} to {new_folder_path}\")\n",
        "            else:\n",
        "                print(f\".fasta file already exists in the destination: {new_fasta_path}\")\n",
        "        else:\n",
        "            print(f\"No matching .fasta file found for {base_name}\")\n",
        "\n",
        "def get_sage_config (json_file_path, peak_folder, static_mods, new_mods, missed_cleavages, min_len, max_len, max_variable_mods, output_config_path):\n",
        "  with open(json_file_path, 'r') as file:\n",
        "    json_data = json.load(file)\n",
        "\n",
        "    peak_files = glob.glob(peak_folder)\n",
        "    print(f\"🗂️ {len(peak_files)} file(s) collected from {peak_folder}\")\n",
        "    json_data['mzml_paths'] = peak_files\n",
        "    json_data['database']['static_mods']=static_mods\n",
        "    json_data['database']['variable_mods'] = new_mods\n",
        "    json_data['database']['enzyme']['missed_cleavages'] = missed_cleavages\n",
        "    json_data['database']['enzyme']['min_len'] = min_len\n",
        "    json_data['database']['enzyme']['max_len'] = max_len\n",
        "    json_data['database']['max_variable_mods']= max_variable_mods\n",
        "    json_data['database']['decoy_tag']=\"rev_\"\n",
        "    json_data['database']['generate_decoys']= True\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "  with open(output_config_path, 'w') as f:\n",
        "    json.dump(json_data, f, indent=4)\n",
        "\n",
        "\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Run Sage\n",
        "import glob\n",
        "import json\n",
        "\n",
        "\n",
        "\n",
        "organise_files (peak_folder)\n",
        "folder_path=peak_folder\n",
        "\n",
        "if use_PTM_plus:\n",
        "  AAs = [AA_1, AA_2, AA_3, AA_4, AA_5]\n",
        "  mods = [AA_1_mod, AA_2_mod, AA_3_mod, AA_4_mod, AA_5_mod]\n",
        "  PTMs = {}\n",
        "  for AA, mod in zip(AAs, mods):\n",
        "    if AA != \"None\":\n",
        "      PTMs[AA] = [mod]\n",
        "  big_folder = glob.glob(f\"{folder_path}/*\")\n",
        "  for folder in big_folder:\n",
        "    if not os.path.isdir(folder):\n",
        "      continue\n",
        "    mzml_files = glob.glob(f\"{folder}/*.{file_type}\")\n",
        "    peak_path=mzml_files[0]\n",
        "    output_json=peak_path.replace(f\".{file_type}\", '.json')\n",
        "    json_file_path = '/content/drive/MyDrive/casanovo/sage/config_general_MQ_fixed_CAM_v1.json'\n",
        "    missed_cleavages=missed_cleavages\n",
        "    min_len = 6\n",
        "    max_len=30\n",
        "    max_variable_mods=max_variable_mods\n",
        "    static_mods={\"C\": 57.021464}\n",
        "    get_sage_config (json_file_path, peak_path, static_mods, PTMs, missed_cleavages, min_len, max_len, max_variable_mods, output_json)\n",
        "    fasta_files=glob.glob(f\"{folder}/*.fasta\")\n",
        "    fasta_path = fasta_files[0]\n",
        "    !sage {output_json} --fasta {fasta_path} --write-pin --output_directory {folder}\n",
        "\n",
        "else:\n",
        "  big_folder = glob.glob(f\"{folder_path}/*\")\n",
        "  for folder in big_folder:\n",
        "    if not os.path.isdir(folder):\n",
        "      continue\n",
        "    mzml_files = glob.glob(f\"{folder}/*.{file_type}\")\n",
        "    peak_path=mzml_files[0]\n",
        "    output_json=peak_path.replace(f\".{file_type}\", '.json')\n",
        "    json_file_path = '/content/drive/MyDrive/casanovo/sage/config_general_MQ_fixed_CAM_v1.json'\n",
        "    missed_cleavages=2\n",
        "    min_len = 6\n",
        "    max_len=30\n",
        "    max_variable_mods=5\n",
        "    static_mods={\"C\": 57.021464}\n",
        "    new_mods={ \"M\": [15.994915],  \"N\": [0.984016], \"Q\": [0.984016]}\n",
        "    get_sage_config (json_file_path, peak_path, static_mods, new_mods, missed_cleavages, min_len, max_len, max_variable_mods, output_json)\n",
        "    fasta_files=glob.glob(f\"{folder}/*.fasta\")\n",
        "    fasta_path = fasta_files[0]\n",
        "    !sage {output_json} --fasta {fasta_path} --write-pin --output_directory {folder}"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "cellView": "form",
        "id": "qDo9CbOsBWIU",
        "outputId": "83c3eb3f-2453-4470-911b-52c2a12964e3"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Created folder: /content/drive/Shareddrives/MS_Palaeome/Odense_RAWs/23_Apr_2024/mzML/test/1007062_BLG_L_T0_1\n",
            "Moved /content/drive/Shareddrives/MS_Palaeome/Odense_RAWs/23_Apr_2024/mzML/test/1007062_BLG_L_T0_1.mzML to /content/drive/Shareddrives/MS_Palaeome/Odense_RAWs/23_Apr_2024/mzML/test/1007062_BLG_L_T0_1\n",
            "Moved /content/drive/Shareddrives/MS_Palaeome/Odense_RAWs/23_Apr_2024/mzML/test/1007062_BLG_L_T0_1_matched.fasta to /content/drive/Shareddrives/MS_Palaeome/Odense_RAWs/23_Apr_2024/mzML/test/1007062_BLG_L_T0_1\n",
            "Created folder: /content/drive/Shareddrives/MS_Palaeome/Odense_RAWs/23_Apr_2024/mzML/test/1007063_BLG_L_T0_2\n",
            "Moved /content/drive/Shareddrives/MS_Palaeome/Odense_RAWs/23_Apr_2024/mzML/test/1007063_BLG_L_T0_2.mzML to /content/drive/Shareddrives/MS_Palaeome/Odense_RAWs/23_Apr_2024/mzML/test/1007063_BLG_L_T0_2\n",
            "Moved /content/drive/Shareddrives/MS_Palaeome/Odense_RAWs/23_Apr_2024/mzML/test/1007063_BLG_L_T0_2_matched.fasta to /content/drive/Shareddrives/MS_Palaeome/Odense_RAWs/23_Apr_2024/mzML/test/1007063_BLG_L_T0_2\n",
            "🗂️ 1 file(s) collected from /content/drive/Shareddrives/MS_Palaeome/Odense_RAWs/23_Apr_2024/mzML/test/1007062_BLG_L_T0_1/1007062_BLG_L_T0_1.mzML\n",
            "\u001b[0m\u001b[38;5;8m[\u001b[0m2024-11-08T09:13:06Z \u001b[0m\u001b[32mINFO \u001b[0m sage\u001b[0m\u001b[38;5;8m]\u001b[0m generated 68804344 fragments, 3353303 peptides in 13672ms\n",
            "\u001b[0m\u001b[38;5;8m[\u001b[0m2024-11-08T09:13:06Z \u001b[0m\u001b[32mINFO \u001b[0m sage\u001b[0m\u001b[38;5;8m]\u001b[0m processing files 0 .. 1 \n",
            "\u001b[0m\u001b[38;5;8m[\u001b[0m2024-11-08T09:13:09Z \u001b[0m\u001b[32mINFO \u001b[0m sage\u001b[0m\u001b[38;5;8m]\u001b[0m - file IO:     2844 ms\n",
            "\u001b[0m\u001b[38;5;8m[\u001b[0m2024-11-08T09:13:11Z \u001b[0m\u001b[32mINFO \u001b[0m sage\u001b[0m\u001b[38;5;8m]\u001b[0m - search:      1555 ms (4513 spectra/s)\n",
            "\u001b[0m\u001b[38;5;8m[\u001b[0m2024-11-08T09:13:11Z \u001b[0m\u001b[32mINFO \u001b[0m sage_core::ml::retention_alignment\u001b[0m\u001b[38;5;8m]\u001b[0m aligning file #0: y = 1.0000x + 0.0000\n",
            "\u001b[0m\u001b[38;5;8m[\u001b[0m2024-11-08T09:13:11Z \u001b[0m\u001b[32mINFO \u001b[0m sage_core::ml::retention_alignment\u001b[0m\u001b[38;5;8m]\u001b[0m aligned retention times across 1 files\n",
            "\u001b[0m\u001b[38;5;8m[\u001b[0m2024-11-08T09:13:11Z \u001b[0m\u001b[32mINFO \u001b[0m sage_core::ml::retention_model\u001b[0m\u001b[38;5;8m]\u001b[0m - fit retention time model, rsq = 0.976836967930682\n",
            "\u001b[0m\u001b[38;5;8m[\u001b[0m2024-11-08T09:13:11Z \u001b[0m\u001b[32mINFO \u001b[0m sage_core::ml::mobility_model\u001b[0m\u001b[38;5;8m]\u001b[0m - fit mobility model, rsq = NaN, mse = 0\n",
            "\u001b[0m\u001b[38;5;8m[\u001b[0m2024-11-08T09:13:11Z \u001b[0m\u001b[32mINFO \u001b[0m sage\u001b[0m\u001b[38;5;8m]\u001b[0m discovered 692 target peptide-spectrum matches at 1% FDR\n",
            "\u001b[0m\u001b[38;5;8m[\u001b[0m2024-11-08T09:13:11Z \u001b[0m\u001b[32mINFO \u001b[0m sage\u001b[0m\u001b[38;5;8m]\u001b[0m discovered 0 target peptides at 1% FDR\n",
            "\u001b[0m\u001b[38;5;8m[\u001b[0m2024-11-08T09:13:11Z \u001b[0m\u001b[32mINFO \u001b[0m sage\u001b[0m\u001b[38;5;8m]\u001b[0m discovered 0 target proteins at 1% FDR\n",
            "{\n",
            "  \"version\": \"0.14.6\",\n",
            "  \"database\": {\n",
            "    \"bucket_size\": 8192,\n",
            "    \"enzyme\": {\n",
            "      \"missed_cleavages\": 2,\n",
            "      \"min_len\": 6,\n",
            "      \"max_len\": 30,\n",
            "      \"cleave_at\": \"KR\",\n",
            "      \"restrict\": null,\n",
            "      \"c_terminal\": null,\n",
            "      \"semi_enzymatic\": null\n",
            "    },\n",
            "    \"fragment_min_mz\": 150.0,\n",
            "    \"fragment_max_mz\": 1500.0,\n",
            "    \"peptide_min_mass\": 500.0,\n",
            "    \"peptide_max_mass\": 5000.0,\n",
            "    \"ion_kinds\": [\n",
            "      \"b\",\n",
            "      \"y\"\n",
            "    ],\n",
            "    \"min_ion_index\": 2,\n",
            "    \"static_mods\": {\n",
            "      \"C\": 57.021465\n",
            "    },\n",
            "    \"variable_mods\": {\n",
            "      \"N\": [\n",
            "        0.984016\n",
            "      ],\n",
            "      \"M\": [\n",
            "        15.9949\n",
            "      ],\n",
            "      \"Q\": [\n",
            "        0.984016\n",
            "      ],\n",
            "      \"P\": [\n",
            "        15.9949\n",
            "      ]\n",
            "    },\n",
            "    \"max_variable_mods\": 3,\n",
            "    \"decoy_tag\": \"rev_\",\n",
            "    \"generate_decoys\": true,\n",
            "    \"fasta\": \"/content/drive/Shareddrives/MS_Palaeome/Odense_RAWs/23_Apr_2024/mzML/test/1007062_BLG_L_T0_1/1007062_BLG_L_T0_1_matched.fasta\"\n",
            "  },\n",
            "  \"quant\": {\n",
            "    \"tmt\": null,\n",
            "    \"tmt_settings\": {\n",
            "      \"level\": 3,\n",
            "      \"sn\": false\n",
            "    },\n",
            "    \"lfq\": false,\n",
            "    \"lfq_settings\": {\n",
            "      \"peak_scoring\": \"Hybrid\",\n",
            "      \"integration\": \"Sum\",\n",
            "      \"spectral_angle\": 0.7,\n",
            "      \"ppm_tolerance\": 5.0,\n",
            "      \"combine_charge_states\": true\n",
            "    }\n",
            "  },\n",
            "  \"precursor_tol\": {\n",
            "    \"ppm\": [\n",
            "      -7.0,\n",
            "      7.0\n",
            "    ]\n",
            "  },\n",
            "  \"fragment_tol\": {\n",
            "    \"ppm\": [\n",
            "      -20.0,\n",
            "      20.0\n",
            "    ]\n",
            "  },\n",
            "  \"precursor_charge\": [\n",
            "    2,\n",
            "    4\n",
            "  ],\n",
            "  \"isotope_errors\": [\n",
            "    0,\n",
            "    1\n",
            "  ],\n",
            "  \"deisotope\": true,\n",
            "  \"chimera\": false,\n",
            "  \"wide_window\": false,\n",
            "  \"min_peaks\": 15,\n",
            "  \"max_peaks\": 150,\n",
            "  \"max_fragment_charge\": 2,\n",
            "  \"min_matched_peaks\": 4,\n",
            "  \"report_psms\": 1,\n",
            "  \"predict_rt\": true,\n",
            "  \"mzml_paths\": [\n",
            "    \"/content/drive/Shareddrives/MS_Palaeome/Odense_RAWs/23_Apr_2024/mzML/test/1007062_BLG_L_T0_1/1007062_BLG_L_T0_1.mzML\"\n",
            "  ],\n",
            "  \"output_paths\": [\n",
            "    \"/content/drive/Shareddrives/MS_Palaeome/Odense_RAWs/23_Apr_2024/mzML/test/1007062_BLG_L_T0_1/results.sage.tsv\",\n",
            "    \"/content/drive/Shareddrives/MS_Palaeome/Odense_RAWs/23_Apr_2024/mzML/test/1007062_BLG_L_T0_1/results.sage.pin\",\n",
            "    \"/content/drive/Shareddrives/MS_Palaeome/Odense_RAWs/23_Apr_2024/mzML/test/1007062_BLG_L_T0_1/results.json\"\n",
            "  ]\n",
            "}\n",
            "\u001b[0m\u001b[38;5;8m[\u001b[0m2024-11-08T09:13:11Z \u001b[0m\u001b[32mINFO \u001b[0m sage\u001b[0m\u001b[38;5;8m]\u001b[0m finished in 18s\n",
            "\u001b[0m\u001b[38;5;8m[\u001b[0m2024-11-08T09:13:11Z \u001b[0m\u001b[32mINFO \u001b[0m sage\u001b[0m\u001b[38;5;8m]\u001b[0m cite: \"Sage: An Open-Source Tool for Fast Proteomics Searching and Quantification at Scale\" https://doi.org/10.1021/acs.jproteome.3c00486\n",
            "🗂️ 1 file(s) collected from /content/drive/Shareddrives/MS_Palaeome/Odense_RAWs/23_Apr_2024/mzML/test/1007063_BLG_L_T0_2/1007063_BLG_L_T0_2.mzML\n",
            "\u001b[0m\u001b[38;5;8m[\u001b[0m2024-11-08T09:13:27Z \u001b[0m\u001b[32mINFO \u001b[0m sage\u001b[0m\u001b[38;5;8m]\u001b[0m generated 63285433 fragments, 3086206 peptides in 13572ms\n",
            "\u001b[0m\u001b[38;5;8m[\u001b[0m2024-11-08T09:13:27Z \u001b[0m\u001b[32mINFO \u001b[0m sage\u001b[0m\u001b[38;5;8m]\u001b[0m processing files 0 .. 1 \n",
            "\u001b[0m\u001b[38;5;8m[\u001b[0m2024-11-08T09:13:31Z \u001b[0m\u001b[32mINFO \u001b[0m sage\u001b[0m\u001b[38;5;8m]\u001b[0m - file IO:     3913 ms\n",
            "\u001b[0m\u001b[38;5;8m[\u001b[0m2024-11-08T09:13:32Z \u001b[0m\u001b[32mINFO \u001b[0m sage\u001b[0m\u001b[38;5;8m]\u001b[0m - search:      1413 ms (5179 spectra/s)\n",
            "\u001b[0m\u001b[38;5;8m[\u001b[0m2024-11-08T09:13:32Z \u001b[0m\u001b[32mINFO \u001b[0m sage_core::ml::retention_alignment\u001b[0m\u001b[38;5;8m]\u001b[0m aligning file #0: y = 1.0000x + 0.0000\n",
            "\u001b[0m\u001b[38;5;8m[\u001b[0m2024-11-08T09:13:32Z \u001b[0m\u001b[32mINFO \u001b[0m sage_core::ml::retention_alignment\u001b[0m\u001b[38;5;8m]\u001b[0m aligned retention times across 1 files\n",
            "\u001b[0m\u001b[38;5;8m[\u001b[0m2024-11-08T09:13:32Z \u001b[0m\u001b[32mINFO \u001b[0m sage_core::ml::retention_model\u001b[0m\u001b[38;5;8m]\u001b[0m - fit retention time model, rsq = 0.9678911847871844\n",
            "\u001b[0m\u001b[38;5;8m[\u001b[0m2024-11-08T09:13:32Z \u001b[0m\u001b[32mINFO \u001b[0m sage_core::ml::mobility_model\u001b[0m\u001b[38;5;8m]\u001b[0m - fit mobility model, rsq = NaN, mse = 0\n",
            "\u001b[0m\u001b[38;5;8m[\u001b[0m2024-11-08T09:13:32Z \u001b[0m\u001b[32mINFO \u001b[0m sage\u001b[0m\u001b[38;5;8m]\u001b[0m discovered 930 target peptide-spectrum matches at 1% FDR\n",
            "\u001b[0m\u001b[38;5;8m[\u001b[0m2024-11-08T09:13:32Z \u001b[0m\u001b[32mINFO \u001b[0m sage\u001b[0m\u001b[38;5;8m]\u001b[0m discovered 118 target peptides at 1% FDR\n",
            "\u001b[0m\u001b[38;5;8m[\u001b[0m2024-11-08T09:13:32Z \u001b[0m\u001b[32mINFO \u001b[0m sage\u001b[0m\u001b[38;5;8m]\u001b[0m discovered 0 target proteins at 1% FDR\n",
            "{\n",
            "  \"version\": \"0.14.6\",\n",
            "  \"database\": {\n",
            "    \"bucket_size\": 8192,\n",
            "    \"enzyme\": {\n",
            "      \"missed_cleavages\": 2,\n",
            "      \"min_len\": 6,\n",
            "      \"max_len\": 30,\n",
            "      \"cleave_at\": \"KR\",\n",
            "      \"restrict\": null,\n",
            "      \"c_terminal\": null,\n",
            "      \"semi_enzymatic\": null\n",
            "    },\n",
            "    \"fragment_min_mz\": 150.0,\n",
            "    \"fragment_max_mz\": 1500.0,\n",
            "    \"peptide_min_mass\": 500.0,\n",
            "    \"peptide_max_mass\": 5000.0,\n",
            "    \"ion_kinds\": [\n",
            "      \"b\",\n",
            "      \"y\"\n",
            "    ],\n",
            "    \"min_ion_index\": 2,\n",
            "    \"static_mods\": {\n",
            "      \"C\": 57.021465\n",
            "    },\n",
            "    \"variable_mods\": {\n",
            "      \"P\": [\n",
            "        15.9949\n",
            "      ],\n",
            "      \"Q\": [\n",
            "        0.984016\n",
            "      ],\n",
            "      \"M\": [\n",
            "        15.9949\n",
            "      ],\n",
            "      \"N\": [\n",
            "        0.984016\n",
            "      ]\n",
            "    },\n",
            "    \"max_variable_mods\": 3,\n",
            "    \"decoy_tag\": \"rev_\",\n",
            "    \"generate_decoys\": true,\n",
            "    \"fasta\": \"/content/drive/Shareddrives/MS_Palaeome/Odense_RAWs/23_Apr_2024/mzML/test/1007063_BLG_L_T0_2/1007063_BLG_L_T0_2_matched.fasta\"\n",
            "  },\n",
            "  \"quant\": {\n",
            "    \"tmt\": null,\n",
            "    \"tmt_settings\": {\n",
            "      \"level\": 3,\n",
            "      \"sn\": false\n",
            "    },\n",
            "    \"lfq\": false,\n",
            "    \"lfq_settings\": {\n",
            "      \"peak_scoring\": \"Hybrid\",\n",
            "      \"integration\": \"Sum\",\n",
            "      \"spectral_angle\": 0.7,\n",
            "      \"ppm_tolerance\": 5.0,\n",
            "      \"combine_charge_states\": true\n",
            "    }\n",
            "  },\n",
            "  \"precursor_tol\": {\n",
            "    \"ppm\": [\n",
            "      -7.0,\n",
            "      7.0\n",
            "    ]\n",
            "  },\n",
            "  \"fragment_tol\": {\n",
            "    \"ppm\": [\n",
            "      -20.0,\n",
            "      20.0\n",
            "    ]\n",
            "  },\n",
            "  \"precursor_charge\": [\n",
            "    2,\n",
            "    4\n",
            "  ],\n",
            "  \"isotope_errors\": [\n",
            "    0,\n",
            "    1\n",
            "  ],\n",
            "  \"deisotope\": true,\n",
            "  \"chimera\": false,\n",
            "  \"wide_window\": false,\n",
            "  \"min_peaks\": 15,\n",
            "  \"max_peaks\": 150,\n",
            "  \"max_fragment_charge\": 2,\n",
            "  \"min_matched_peaks\": 4,\n",
            "  \"report_psms\": 1,\n",
            "  \"predict_rt\": true,\n",
            "  \"mzml_paths\": [\n",
            "    \"/content/drive/Shareddrives/MS_Palaeome/Odense_RAWs/23_Apr_2024/mzML/test/1007063_BLG_L_T0_2/1007063_BLG_L_T0_2.mzML\"\n",
            "  ],\n",
            "  \"output_paths\": [\n",
            "    \"/content/drive/Shareddrives/MS_Palaeome/Odense_RAWs/23_Apr_2024/mzML/test/1007063_BLG_L_T0_2/results.sage.tsv\",\n",
            "    \"/content/drive/Shareddrives/MS_Palaeome/Odense_RAWs/23_Apr_2024/mzML/test/1007063_BLG_L_T0_2/results.sage.pin\",\n",
            "    \"/content/drive/Shareddrives/MS_Palaeome/Odense_RAWs/23_Apr_2024/mzML/test/1007063_BLG_L_T0_2/results.json\"\n",
            "  ]\n",
            "}\n",
            "\u001b[0m\u001b[38;5;8m[\u001b[0m2024-11-08T09:13:32Z \u001b[0m\u001b[32mINFO \u001b[0m sage\u001b[0m\u001b[38;5;8m]\u001b[0m finished in 19s\n",
            "\u001b[0m\u001b[38;5;8m[\u001b[0m2024-11-08T09:13:32Z \u001b[0m\u001b[32mINFO \u001b[0m sage\u001b[0m\u001b[38;5;8m]\u001b[0m cite: \"Sage: An Open-Source Tool for Fast Proteomics Searching and Quantification at Scale\" https://doi.org/10.1021/acs.jproteome.3c00486\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Brew Mokapot\n",
        "\n",
        "import mokapot\n",
        "from xgboost import XGBClassifier\n",
        "from sklearn.neural_network import MLPClassifier\n",
        "from sklearn.model_selection import GridSearchCV\n",
        "import numpy as np\n",
        "\n",
        "'''\n",
        "XGBoost schema from Fondrie & Noble 2021). A non-linear XGBoost seems to be better for rescoring open search results.\n",
        "\n",
        "'''\n",
        "\n",
        "grid = {\n",
        "            \"scale_pos_weight\": np.logspace(0, 2, 3),\n",
        "            \"max_depth\": [1, 3, 6],\n",
        "            \"min_child_weight\": [1, 10, 100],\n",
        "            \"gamma\": [0, 1, 10],\n",
        "        }\n",
        "xgb_mod = GridSearchCV(\n",
        "            XGBClassifier(),\n",
        "            param_grid=grid,\n",
        "            n_jobs=1,\n",
        "            cv=3,\n",
        "            scoring=\"roc_auc\",\n",
        "        )\n",
        "\n",
        "def get_all_pin_files(folder_path):\n",
        "    psm_files = []\n",
        "    for root, dirs, files in os.walk(folder_path):\n",
        "        for file in files:\n",
        "            if file.endswith('.pin'):\n",
        "                full_path = os.path.join(root, file)\n",
        "                psm_files.append(full_path)\n",
        "    return psm_files\n",
        "\n",
        "if use_Mokapot:\n",
        "\n",
        " folder_path=peak_folder\n",
        "\n",
        " if joint_modelling:\n",
        "  psm_files=get_all_pin_files(peak_folder)\n",
        "  if default_Percolator:\n",
        "    svm=mokapot.PercolatorModel()\n",
        "    psm_list = mokapot.read_pin(psm_files)\n",
        "    results, models = mokapot.brew(psm_list, svm)\n",
        "    result_files = results.to_txt(peak_folder)\n",
        "  else:\n",
        "    mod = mokapot.Model(xgb_mod)\n",
        "    psm_list = mokapot.read_pin(psm_files)\n",
        "    results, models = mokapot.brew(psm_list, mod)\n",
        "    result_files = results.to_txt(peak_folder)\n",
        "\n",
        " else:\n",
        "  big_folder = sorted(glob.glob(f\"{folder_path}/*\"))\n",
        "  for folder in big_folder:\n",
        "    if not os.path.isdir(folder):\n",
        "      continue\n",
        "    print(f\"Processing folder: {folder}\")\n",
        "    pin_files = glob.glob(f\"{folder}/*.pin\")\n",
        "    if not pin_files:\n",
        "      print(f\"No .pin files found in {folder}. Skipping...\")\n",
        "      continue\n",
        "    pin=pin_files[0]\n",
        "    if default_Percolator:\n",
        "      svm=mokapot.PercolatorModel()\n",
        "      psm_list = mokapot.read_pin(pin)\n",
        "      results, models = mokapot.brew(psm_list, svm)\n",
        "      result_files = results.to_txt(folder)\n",
        "    else:\n",
        "      mod = mokapot.Model(xgb_mod)\n",
        "      psm_list = mokapot.read_pin(pin)\n",
        "      results, models = mokapot.brew(psm_list, mod)\n",
        "      result_files = results.to_txt(folder)\n",
        "\n",
        "else:\n",
        "  print(\"Mokapot not brewed\")\n",
        "\n",
        "\n"
      ],
      "metadata": {
        "id": "vGSAYJN1_923",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "cellView": "form",
        "outputId": "1a3662e3-ca81-4459-f7c5-93f7c795225a"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Processing folder: /content/drive/Shareddrives/MS_Palaeome/Odense_RAWs/23_Apr_2024/mzML/test/1007062_BLG_L_T0_1\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/usr/local/lib/python3.10/dist-packages/mokapot/parsers/pin.py:238: FutureWarning: errors='ignore' is deprecated and will raise in a future version. Use to_numeric without passing `errors` and catch exceptions explicitly instead\n",
            "  yield psms.apply(pd.to_numeric, errors=\"ignore\")\n",
            "/usr/local/lib/python3.10/dist-packages/numpy/ma/core.py:2820: RuntimeWarning: invalid value encountered in cast\n",
            "  _data = np.array(data, dtype=dtype, copy=copy,\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Warning: IRLS did not converge with maxIter = 50\n",
            "Warning: IRLS did not converge with maxIter = 50\n",
            "Warning: IRLS did not converge with maxIter = 50\n",
            "Warning: IRLS did not converge with maxIter = 50\n",
            "Warning: IRLS did not converge with maxIter = 50\n",
            "Warning: IRLS did not converge with maxIter = 50\n",
            "Warning: IRLS did not converge with maxIter = 50\n",
            "Warning: IRLS did not converge with maxIter = 50\n",
            "Warning: IRLS did not converge with maxIter = 50\n",
            "Warning: IRLS did not converge with maxIter = 50\n",
            "Warning: IRLS did not converge with maxIter = 50\n",
            "Warning: IRLS did not converge with maxIter = 50\n",
            "Warning: IRLS did not converge with maxIter = 50\n",
            "Processing folder: /content/drive/Shareddrives/MS_Palaeome/Odense_RAWs/23_Apr_2024/mzML/test/1007063_BLG_L_T0_2\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/usr/local/lib/python3.10/dist-packages/mokapot/parsers/pin.py:238: FutureWarning: errors='ignore' is deprecated and will raise in a future version. Use to_numeric without passing `errors` and catch exceptions explicitly instead\n",
            "  yield psms.apply(pd.to_numeric, errors=\"ignore\")\n"
          ]
        }
      ]
    }
  ]
}
