{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "colab_type": "text",
    "id": "view-in-github"
   },
   "source": [
    "<a href=\"https://colab.research.google.com/github/yc386/orthrus_metaproteomics/blob/main/annotated_orthrus_pt1_v1.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "VpisBTGyhi5v"
   },
   "source": [
    "<img src='https://drive.google.com/uc?export=view&id=19rmmQI1H2nIqgU598WROTcUNhOUoXcBP' width='400px' align='right'>\n",
    "\n",
    "# **Readme**\n",
    "\n",
    "---\n",
    "Orthrus 🐾 is a hybrid, two-software pipeline that integrates [Casanovo](https://github.com/Noble-Lab/casanovo) (an AI transformer) with [Sage](https://github.com/lazear/sage) (a fast database search engine with advanced features like retention time alignment and machine learning-based rescoring).\n",
    "\n",
    "Designed to handle large search space and difficulties of selecting databases in metaproteomics and palaeoproteomics, Orthrus leverages de novo sequencing to define sample-specific databases, and uses probability ranking and conventional database searching to control FDRs (false discovery rates).\n",
    "\n",
    "Orthrus can be run online using Google Colab 🥳, or locally via Anaconda 🐍.\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "Fva_FXk0htYl"
   },
   "source": [
    "# **Please note**❗️\n",
    "*   Before walking the dog, please change the runtime type to GPU (A100, L4, or T4. A100 most efficient but T4 is free)\n",
    "*   If you would like to connect your Google Drive, click the folder image 🗂️ on the left and mount the drive.\n",
    "*  Click `File` (top left) and save a copy in Drive or Github"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "-K0wYqUruZmx"
   },
   "source": [
    "# Run `Orthrus`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "Cv26MhBMhbT5"
   },
   "outputs": [],
   "source": [
    "#@title Add inputs -> click `Runtime` -> `Run all`\n",
    "#@markdown **Casanovo inputs**\n",
    "folder_path=\"/content/drive/Shareddrives/MS_Palaeome/Odense_RAWs/23_Apr_2024/mzML/test\"#@param {type:\"string\"}\n",
    "#@markdown - a folder contains single or multiple `.mzML` or `.mgf` files for `Casanovo`. Please check only _ (underscore) and no other special characters or space in a file name.\n",
    "file_type=\"mzML\" #@param [\"mzML\", \"mgf\"]\n",
    "#@markdown - use the drop-down menu to choose the instrument file type\n",
    "\n",
    "use_default = True #@param {type:\"boolean\"}\n",
    "#@markdown **Advanced Options (ignored if using default settings)**\n",
    "\n",
    "model = \"/content/drive/MyDrive/casanovo/models/casanovo4/ckpt/casanovo_massivekb_worker1.ckpt\" #@param {type:\"string\"}\n",
    "#@markdown - a model for Casanovo prediction `.ckpt`\n",
    "config = \"/content/drive/MyDrive/casanovo/models/casanovo4/yaml/config_420.yaml\" #@param {type:\"string\"}\n",
    "#@markdown - a `.yaml` configuration file for Casanovo\n",
    "\n",
    "#@markdown **Inputs for converting Casanovo results to a `.fasta`**\n",
    "use_SwissProt = True #@param {type:\"boolean\"}\n",
    "#@markdown - use the latest, reviewed SwissProt form the UniProt FTP\n",
    "database_path=\"\"#@param {type:\"string\"}\n",
    "#@markdown - path to a database (`.fasta`) for shortlisting proteins based on de novo results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "F9pGGlDUimMo",
    "outputId": "a8e81df3-8963-4180-d01e-6e99b37cbf75"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting casanovo\n",
      "  Downloading casanovo-4.2.1-py3-none-any.whl.metadata (2.8 kB)\n",
      "Collecting biopython\n",
      "  Downloading biopython-1.84-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (12 kB)\n",
      "Collecting mokapot\n",
      "  Downloading mokapot-0.10.0-py3-none-any.whl.metadata (4.5 kB)\n",
      "Collecting appdirs (from casanovo)\n",
      "  Downloading appdirs-1.4.4-py2.py3-none-any.whl.metadata (9.0 kB)\n",
      "Collecting lightning>=2.0 (from casanovo)\n",
      "  Downloading lightning-2.4.0-py3-none-any.whl.metadata (38 kB)\n",
      "Requirement already satisfied: click in /usr/local/lib/python3.10/dist-packages (from casanovo) (8.1.7)\n",
      "Collecting depthcharge-ms<0.3.0,>=0.2.3 (from casanovo)\n",
      "  Downloading depthcharge_ms-0.2.3-py3-none-any.whl.metadata (4.7 kB)\n",
      "Requirement already satisfied: natsort in /usr/local/lib/python3.10/dist-packages (from casanovo) (8.4.0)\n",
      "Requirement already satisfied: numpy<2.0 in /usr/local/lib/python3.10/dist-packages (from casanovo) (1.26.4)\n",
      "Requirement already satisfied: pandas in /usr/local/lib/python3.10/dist-packages (from casanovo) (2.2.2)\n",
      "Requirement already satisfied: psutil in /usr/local/lib/python3.10/dist-packages (from casanovo) (5.9.5)\n",
      "Collecting PyGithub (from casanovo)\n",
      "  Downloading PyGithub-2.5.0-py3-none-any.whl.metadata (3.9 kB)\n",
      "Requirement already satisfied: PyYAML in /usr/local/lib/python3.10/dist-packages (from casanovo) (6.0.2)\n",
      "Requirement already satisfied: requests in /usr/local/lib/python3.10/dist-packages (from casanovo) (2.32.3)\n",
      "Collecting rich-click>=1.6.1 (from casanovo)\n",
      "  Downloading rich_click-1.8.3-py3-none-any.whl.metadata (7.9 kB)\n",
      "Requirement already satisfied: scikit-learn in /usr/local/lib/python3.10/dist-packages (from casanovo) (1.5.2)\n",
      "Collecting spectrum-utils (from casanovo)\n",
      "  Downloading spectrum_utils-0.4.2-py3-none-any.whl.metadata (3.9 kB)\n",
      "Requirement already satisfied: tensorboard in /usr/local/lib/python3.10/dist-packages (from casanovo) (2.17.0)\n",
      "Requirement already satisfied: torch>=2.1 in /usr/local/lib/python3.10/dist-packages (from casanovo) (2.5.0+cu121)\n",
      "Requirement already satisfied: tqdm in /usr/local/lib/python3.10/dist-packages (from casanovo) (4.66.6)\n",
      "Requirement already satisfied: numba>=0.48.0 in /usr/local/lib/python3.10/dist-packages (from mokapot) (0.60.0)\n",
      "Requirement already satisfied: matplotlib>=3.1.3 in /usr/local/lib/python3.10/dist-packages (from mokapot) (3.8.0)\n",
      "Requirement already satisfied: lxml>=4.6.2 in /usr/local/lib/python3.10/dist-packages (from mokapot) (5.3.0)\n",
      "Collecting triqler>=0.6.2 (from mokapot)\n",
      "  Downloading triqler-0.7.3-py3-none-any.whl.metadata (11 kB)\n",
      "Requirement already satisfied: joblib>=1.1.0 in /usr/local/lib/python3.10/dist-packages (from mokapot) (1.4.2)\n",
      "Requirement already satisfied: importlib-metadata>=5.1.0 in /usr/local/lib/python3.10/dist-packages (from mokapot) (8.5.0)\n",
      "Collecting pyteomics>=4.4.2 (from depthcharge-ms<0.3.0,>=0.2.3->casanovo)\n",
      "  Downloading pyteomics-4.7.5-py3-none-any.whl.metadata (6.5 kB)\n",
      "Requirement already satisfied: h5py>=3.7.0 in /usr/local/lib/python3.10/dist-packages (from depthcharge-ms<0.3.0,>=0.2.3->casanovo) (3.12.1)\n",
      "Requirement already satisfied: einops>=0.4.1 in /usr/local/lib/python3.10/dist-packages (from depthcharge-ms<0.3.0,>=0.2.3->casanovo) (0.8.0)\n",
      "Requirement already satisfied: zipp>=3.20 in /usr/local/lib/python3.10/dist-packages (from importlib-metadata>=5.1.0->mokapot) (3.20.2)\n",
      "Requirement already satisfied: fsspec<2026.0,>=2022.5.0 in /usr/local/lib/python3.10/dist-packages (from fsspec[http]<2026.0,>=2022.5.0->lightning>=2.0->casanovo) (2024.10.0)\n",
      "Collecting lightning-utilities<2.0,>=0.10.0 (from lightning>=2.0->casanovo)\n",
      "  Downloading lightning_utilities-0.11.8-py3-none-any.whl.metadata (5.2 kB)\n",
      "Requirement already satisfied: packaging<25.0,>=20.0 in /usr/local/lib/python3.10/dist-packages (from lightning>=2.0->casanovo) (24.1)\n",
      "Collecting torchmetrics<3.0,>=0.7.0 (from lightning>=2.0->casanovo)\n",
      "  Downloading torchmetrics-1.5.1-py3-none-any.whl.metadata (20 kB)\n",
      "Requirement already satisfied: typing-extensions<6.0,>=4.4.0 in /usr/local/lib/python3.10/dist-packages (from lightning>=2.0->casanovo) (4.12.2)\n",
      "Collecting pytorch-lightning (from lightning>=2.0->casanovo)\n",
      "  Downloading pytorch_lightning-2.4.0-py3-none-any.whl.metadata (21 kB)\n",
      "Requirement already satisfied: contourpy>=1.0.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=3.1.3->mokapot) (1.3.0)\n",
      "Requirement already satisfied: cycler>=0.10 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=3.1.3->mokapot) (0.12.1)\n",
      "Requirement already satisfied: fonttools>=4.22.0 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=3.1.3->mokapot) (4.54.1)\n",
      "Requirement already satisfied: kiwisolver>=1.0.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=3.1.3->mokapot) (1.4.7)\n",
      "Requirement already satisfied: pillow>=6.2.0 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=3.1.3->mokapot) (10.4.0)\n",
      "Requirement already satisfied: pyparsing>=2.3.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=3.1.3->mokapot) (3.2.0)\n",
      "Requirement already satisfied: python-dateutil>=2.7 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=3.1.3->mokapot) (2.8.2)\n",
      "Requirement already satisfied: llvmlite<0.44,>=0.43.0dev0 in /usr/local/lib/python3.10/dist-packages (from numba>=0.48.0->mokapot) (0.43.0)\n",
      "Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.10/dist-packages (from pandas->casanovo) (2024.2)\n",
      "Requirement already satisfied: tzdata>=2022.7 in /usr/local/lib/python3.10/dist-packages (from pandas->casanovo) (2024.2)\n",
      "Requirement already satisfied: rich>=10.7 in /usr/local/lib/python3.10/dist-packages (from rich-click>=1.6.1->casanovo) (13.9.3)\n",
      "Requirement already satisfied: scipy>=1.6.0 in /usr/local/lib/python3.10/dist-packages (from scikit-learn->casanovo) (1.13.1)\n",
      "Requirement already satisfied: threadpoolctl>=3.1.0 in /usr/local/lib/python3.10/dist-packages (from scikit-learn->casanovo) (3.5.0)\n",
      "Requirement already satisfied: filelock in /usr/local/lib/python3.10/dist-packages (from torch>=2.1->casanovo) (3.16.1)\n",
      "Requirement already satisfied: networkx in /usr/local/lib/python3.10/dist-packages (from torch>=2.1->casanovo) (3.4.2)\n",
      "Requirement already satisfied: jinja2 in /usr/local/lib/python3.10/dist-packages (from torch>=2.1->casanovo) (3.1.4)\n",
      "Requirement already satisfied: sympy==1.13.1 in /usr/local/lib/python3.10/dist-packages (from torch>=2.1->casanovo) (1.13.1)\n",
      "Requirement already satisfied: mpmath<1.4,>=1.1.0 in /usr/local/lib/python3.10/dist-packages (from sympy==1.13.1->torch>=2.1->casanovo) (1.3.0)\n",
      "Collecting job-pool<0.3.0,>=0.2.6 (from triqler>=0.6.2->mokapot)\n",
      "  Downloading job_pool-0.2.6-py3-none-any.whl.metadata (2.1 kB)\n",
      "Collecting pynacl>=1.4.0 (from PyGithub->casanovo)\n",
      "  Downloading PyNaCl-1.5.0-cp36-abi3-manylinux_2_17_x86_64.manylinux2014_x86_64.manylinux_2_24_x86_64.whl.metadata (8.6 kB)\n",
      "Requirement already satisfied: pyjwt>=2.4.0 in /usr/local/lib/python3.10/dist-packages (from pyjwt[crypto]>=2.4.0->PyGithub->casanovo) (2.9.0)\n",
      "Requirement already satisfied: urllib3>=1.26.0 in /usr/local/lib/python3.10/dist-packages (from PyGithub->casanovo) (2.2.3)\n",
      "Requirement already satisfied: Deprecated in /usr/local/lib/python3.10/dist-packages (from PyGithub->casanovo) (1.2.14)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.10/dist-packages (from requests->casanovo) (3.4.0)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.10/dist-packages (from requests->casanovo) (3.10)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.10/dist-packages (from requests->casanovo) (2024.8.30)\n",
      "Collecting fastobo (from spectrum-utils->casanovo)\n",
      "  Downloading fastobo-0.12.3-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (6.3 kB)\n",
      "Collecting lark>=1.0 (from spectrum-utils->casanovo)\n",
      "  Downloading lark-1.2.2-py3-none-any.whl.metadata (1.8 kB)\n",
      "Requirement already satisfied: absl-py>=0.4 in /usr/local/lib/python3.10/dist-packages (from tensorboard->casanovo) (1.4.0)\n",
      "Requirement already satisfied: grpcio>=1.48.2 in /usr/local/lib/python3.10/dist-packages (from tensorboard->casanovo) (1.64.1)\n",
      "Requirement already satisfied: markdown>=2.6.8 in /usr/local/lib/python3.10/dist-packages (from tensorboard->casanovo) (3.7)\n",
      "Requirement already satisfied: protobuf!=4.24.0,<5.0.0,>=3.19.6 in /usr/local/lib/python3.10/dist-packages (from tensorboard->casanovo) (3.20.3)\n",
      "Requirement already satisfied: setuptools>=41.0.0 in /usr/local/lib/python3.10/dist-packages (from tensorboard->casanovo) (75.1.0)\n",
      "Requirement already satisfied: six>1.9 in /usr/local/lib/python3.10/dist-packages (from tensorboard->casanovo) (1.16.0)\n",
      "Requirement already satisfied: tensorboard-data-server<0.8.0,>=0.7.0 in /usr/local/lib/python3.10/dist-packages (from tensorboard->casanovo) (0.7.2)\n",
      "Requirement already satisfied: werkzeug>=1.0.1 in /usr/local/lib/python3.10/dist-packages (from tensorboard->casanovo) (3.0.6)\n",
      "Requirement already satisfied: aiohttp!=4.0.0a0,!=4.0.0a1 in /usr/local/lib/python3.10/dist-packages (from fsspec[http]<2026.0,>=2022.5.0->lightning>=2.0->casanovo) (3.10.10)\n",
      "Requirement already satisfied: cryptography>=3.4.0 in /usr/local/lib/python3.10/dist-packages (from pyjwt[crypto]>=2.4.0->PyGithub->casanovo) (43.0.3)\n",
      "Requirement already satisfied: cffi>=1.4.1 in /usr/local/lib/python3.10/dist-packages (from pynacl>=1.4.0->PyGithub->casanovo) (1.17.1)\n",
      "Requirement already satisfied: markdown-it-py>=2.2.0 in /usr/local/lib/python3.10/dist-packages (from rich>=10.7->rich-click>=1.6.1->casanovo) (3.0.0)\n",
      "Requirement already satisfied: pygments<3.0.0,>=2.13.0 in /usr/local/lib/python3.10/dist-packages (from rich>=10.7->rich-click>=1.6.1->casanovo) (2.18.0)\n",
      "Requirement already satisfied: MarkupSafe>=2.1.1 in /usr/local/lib/python3.10/dist-packages (from werkzeug>=1.0.1->tensorboard->casanovo) (3.0.2)\n",
      "Requirement already satisfied: wrapt<2,>=1.10 in /usr/local/lib/python3.10/dist-packages (from Deprecated->PyGithub->casanovo) (1.16.0)\n",
      "Requirement already satisfied: aiohappyeyeballs>=2.3.0 in /usr/local/lib/python3.10/dist-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->fsspec[http]<2026.0,>=2022.5.0->lightning>=2.0->casanovo) (2.4.3)\n",
      "Requirement already satisfied: aiosignal>=1.1.2 in /usr/local/lib/python3.10/dist-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->fsspec[http]<2026.0,>=2022.5.0->lightning>=2.0->casanovo) (1.3.1)\n",
      "Requirement already satisfied: attrs>=17.3.0 in /usr/local/lib/python3.10/dist-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->fsspec[http]<2026.0,>=2022.5.0->lightning>=2.0->casanovo) (24.2.0)\n",
      "Requirement already satisfied: frozenlist>=1.1.1 in /usr/local/lib/python3.10/dist-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->fsspec[http]<2026.0,>=2022.5.0->lightning>=2.0->casanovo) (1.5.0)\n",
      "Requirement already satisfied: multidict<7.0,>=4.5 in /usr/local/lib/python3.10/dist-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->fsspec[http]<2026.0,>=2022.5.0->lightning>=2.0->casanovo) (6.1.0)\n",
      "Requirement already satisfied: yarl<2.0,>=1.12.0 in /usr/local/lib/python3.10/dist-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->fsspec[http]<2026.0,>=2022.5.0->lightning>=2.0->casanovo) (1.17.0)\n",
      "Requirement already satisfied: async-timeout<5.0,>=4.0 in /usr/local/lib/python3.10/dist-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->fsspec[http]<2026.0,>=2022.5.0->lightning>=2.0->casanovo) (4.0.3)\n",
      "Requirement already satisfied: pycparser in /usr/local/lib/python3.10/dist-packages (from cffi>=1.4.1->pynacl>=1.4.0->PyGithub->casanovo) (2.22)\n",
      "Requirement already satisfied: mdurl~=0.1 in /usr/local/lib/python3.10/dist-packages (from markdown-it-py>=2.2.0->rich>=10.7->rich-click>=1.6.1->casanovo) (0.1.2)\n",
      "Requirement already satisfied: propcache>=0.2.0 in /usr/local/lib/python3.10/dist-packages (from yarl<2.0,>=1.12.0->aiohttp!=4.0.0a0,!=4.0.0a1->fsspec[http]<2026.0,>=2022.5.0->lightning>=2.0->casanovo) (0.2.0)\n",
      "Downloading casanovo-4.2.1-py3-none-any.whl (58 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m58.3/58.3 kB\u001b[0m \u001b[31m5.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading biopython-1.84-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.2 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.2/3.2 MB\u001b[0m \u001b[31m86.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading mokapot-0.10.0-py3-none-any.whl (74 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m74.4/74.4 kB\u001b[0m \u001b[31m6.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading depthcharge_ms-0.2.3-py3-none-any.whl (27 kB)\n",
      "Downloading lightning-2.4.0-py3-none-any.whl (810 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m811.0/811.0 kB\u001b[0m \u001b[31m45.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading rich_click-1.8.3-py3-none-any.whl (35 kB)\n",
      "Downloading triqler-0.7.3-py3-none-any.whl (64 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m64.2/64.2 kB\u001b[0m \u001b[31m5.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading appdirs-1.4.4-py2.py3-none-any.whl (9.6 kB)\n",
      "Downloading PyGithub-2.5.0-py3-none-any.whl (375 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m375.9/375.9 kB\u001b[0m \u001b[31m31.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading spectrum_utils-0.4.2-py3-none-any.whl (45 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m45.5/45.5 kB\u001b[0m \u001b[31m3.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading job_pool-0.2.6-py3-none-any.whl (9.1 kB)\n",
      "Downloading lark-1.2.2-py3-none-any.whl (111 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m111.0/111.0 kB\u001b[0m \u001b[31m9.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading lightning_utilities-0.11.8-py3-none-any.whl (26 kB)\n",
      "Downloading PyNaCl-1.5.0-cp36-abi3-manylinux_2_17_x86_64.manylinux2014_x86_64.manylinux_2_24_x86_64.whl (856 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m856.7/856.7 kB\u001b[0m \u001b[31m53.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading pyteomics-4.7.5-py3-none-any.whl (238 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m239.0/239.0 kB\u001b[0m \u001b[31m21.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading torchmetrics-1.5.1-py3-none-any.whl (890 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m890.6/890.6 kB\u001b[0m \u001b[31m57.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading fastobo-0.12.3-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (2.0 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m2.0/2.0 MB\u001b[0m \u001b[31m28.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading pytorch_lightning-2.4.0-py3-none-any.whl (815 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m815.2/815.2 kB\u001b[0m \u001b[31m54.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hInstalling collected packages: pyteomics, appdirs, lightning-utilities, lark, job-pool, fastobo, biopython, pynacl, triqler, torchmetrics, spectrum-utils, rich-click, depthcharge-ms, PyGithub, mokapot, pytorch-lightning, lightning, casanovo\n",
      "Successfully installed PyGithub-2.5.0 appdirs-1.4.4 biopython-1.84 casanovo-4.2.1 depthcharge-ms-0.2.3 fastobo-0.12.3 job-pool-0.2.6 lark-1.2.2 lightning-2.4.0 lightning-utilities-0.11.8 mokapot-0.10.0 pynacl-1.5.0 pyteomics-4.7.5 pytorch-lightning-2.4.0 rich-click-1.8.3 spectrum-utils-0.4.2 torchmetrics-1.5.1 triqler-0.7.3\n",
      "installing conda...\n",
      "uniprot_sprot.fasta.gz downloaded successfully.\n",
      "uniprot_sprot.fasta decompressed successfully.\n",
      "CPU times: user 5.17 s, sys: 611 ms, total: 5.78 s\n",
      "Wall time: 1min 33s\n"
     ]
    }
   ],
   "source": [
    "#@title install dependencies & modules\n",
    "\n",
    "%%time\n",
    "\n",
    "!pip install casanovo biopython mokapot\n",
    "\n",
    "import os\n",
    "\n",
    "if not os.path.isfile(\"Orthrus_READY\"):\n",
    "  print(\"installing conda...\")\n",
    "  os.system(\"wget -qnc https://github.com/conda-forge/miniforge/releases/latest/download/Mambaforge-Linux-x86_64.sh\")\n",
    "  os.system(\"bash Mambaforge-Linux-x86_64.sh -bfp /usr/local\")\n",
    "  os.system(\"touch Orthrus_READY\")\n",
    "  os.system(f\"conda install -c bioconda -c conda-forge sage-proteomics -y -q\")\n",
    "\n",
    "\n",
    "import glob\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n",
    "import json\n",
    "from pyteomics import mztab\n",
    "from numpy import string_\n",
    "from joblib import Parallel, delayed\n",
    "from itertools import chain\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.naive_bayes import GaussianNB\n",
    "from sklearn.preprocessing import MinMaxScaler\n",
    "from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio import SeqIO\n",
    "import requests\n",
    "import gzip\n",
    "import shutil\n",
    "\n",
    "'''\n",
    "parse a .mztab file using pyteomics\n",
    "add naked sequences (without PTMs) & the sequence length\n",
    "input=path to .mztab file\n",
    "output=pandas dataframe\n",
    "\n",
    "'''\n",
    "pattern = re.compile(r'(.\\d*\\.?\\d+)')\n",
    "\n",
    "def prep_mztab(mztab_path):\n",
    "  m = mztab.MzTab(mztab_path)\n",
    "  df = m.spectrum_match_table\n",
    "  if df is None or df.empty:\n",
    "    raise ValueError(f\"{mztab_path} is empty\")\n",
    "  if 'sequence' not in df.columns:\n",
    "    raise KeyError(f\"'sequence' column is missing in the file: {mztab_path}\")\n",
    "  df.reset_index(drop=True)\n",
    "  df1 = df.assign(sequence_naked=df['sequence'].str.replace(pattern, '', regex=True))\n",
    "  df2= df1.assign(nAA=df1['sequence_naked'].str.len())\n",
    "  df3=df2.sort_values(by='sequence_naked').drop_duplicates(subset='sequence_naked', keep=\"first\").reset_index(drop=True)\n",
    "  return df3\n",
    "\n",
    "'''\n",
    "parse a .fasta file using biopython\n",
    "add UniProt ID e.g. P02754\n",
    "input=path to .fasta file\n",
    "output=pandas dataframe\n",
    "\n",
    "'''\n",
    "\n",
    "def fasta_to_df(fasta_file):\n",
    "\n",
    "  data = []\n",
    "\n",
    "  for record in SeqIO.parse(fasta_file, \"fasta\"):\n",
    "\n",
    "    protein_id = record.id\n",
    "    description = record.description\n",
    "    sequence = str(record.seq)\n",
    "    if not sequence:\n",
    "      raise ValueError(f\"Record with ID '{protein_id}' has no sequence in the fasta file.\")\n",
    "\n",
    "    data.append((protein_id, description, sequence))\n",
    "\n",
    "  df = pd.DataFrame(data, columns=[\"Protein_ID\", \"Description\", \"Sequence\"])\n",
    "  df1=df.assign(UniProt_ID=df['Protein_ID'].str.split('|').str[1])\n",
    "\n",
    "  return df1\n",
    "\n",
    "\n",
    "'''\n",
    "filter a Casanovo output file based on the maximum value below 0\n",
    "search_engine_score[1] is a score assigned to each prediction by Casanovo, max=1,\n",
    "if negative then outside the mass tolerance\n",
    "input=pandas dataframe\n",
    "output=pandas dataframe\n",
    "\n",
    "'''\n",
    "\n",
    "def casa_filter (df):\n",
    "  np_array = df['search_engine_score[1]'].to_numpy()\n",
    "  max_below_zero = np_array[np_array < 0].max()\n",
    "  df1=df[df['search_engine_score[1]']>=max_below_zero]\n",
    "  return df1\n",
    "\n",
    "\n",
    "#prepare overlapping sequence tags for string matching\n",
    "def get_seq_tags (sequence, k):\n",
    "  return set(sequence[i:i+k] for i in range(len(sequence) - k + 1))\n",
    "\n",
    "\n",
    "'''\n",
    "match de novo-based tags with database tags\n",
    "I=L in a reference database\n",
    "inputs=path to fasta, filtered casanovo output dataframe, tag size=k, chunk size=10000 for processing\n",
    "output=pandas dataframe\n",
    "\n",
    "'''\n",
    "\n",
    "def matching_count_v5 (df, df1, k, chunk_size=10000):\n",
    "\n",
    "  sequence_set = get_seq_tags(''.join(chain.from_iterable(df1['sequence_naked'].astype(str))), k)\n",
    "  print(f\"📝 {len(sequence_set)} tags regenerated. Starting matching...\")\n",
    "  result_df = pd.DataFrame()\n",
    "  for start in range(0, len(df), chunk_size):\n",
    "    chunk = df.iloc[start:start+chunk_size].copy()\n",
    "    chunk['seq_tags'] = chunk['Sequence'].astype(str).str.replace('I', 'L').apply(lambda x: get_seq_tags(x, k))\n",
    "    chunk['matched_count'] = chunk['seq_tags'].apply(lambda seq_tags: len(seq_tags & sequence_set))\n",
    "    chunk = chunk.assign(matched=chunk['matched_count'].apply(lambda x: 1 if x >= 2 else 0))\n",
    "    result_df = pd.concat([result_df, chunk], ignore_index=True)\n",
    "  total_matches=result_df['matched_count'].sum()\n",
    "  print(f\"Completed! {total_matches} matched 👍 \")\n",
    "  return result_df\n",
    "\n",
    "#get tryptic peptides per database entry\n",
    "def count_tryptic_peptides(sequence):\n",
    "  pattern=r'(?<=[KR])'\n",
    "\n",
    "  peptides = re.split(pattern, sequence)\n",
    "\n",
    "  filtered_peptides = [peptide for peptide in peptides if len(peptide) >= 6]\n",
    "\n",
    "  return len(filtered_peptides)\n",
    "\n",
    "#prepare a dataframe for NB classification\n",
    "def prep_Bayes (df):\n",
    "  print('🧑‍💻 Start Bayes probabilistic ranking...')\n",
    "  df1=df.assign(length=df['Sequence'].astype(str).str.len(),\n",
    "                 tryptic_count=df['Sequence'].apply(count_tryptic_peptides),\n",
    "                 tag_count=df['seq_tags'].apply(len))\n",
    "  df2=df1.assign(SAF=df1['matched_count']/df1['length'],\n",
    "                 try_ratio=df1['tryptic_count']/df1['tag_count']\n",
    "                 )\n",
    "  return df2\n",
    "\n",
    "\n",
    "'''\n",
    "ranking matched proteins based on SAF and try_ratio\n",
    "values normalised before NB classification\n",
    "most probable matches (>= 95%) are shortlised\n",
    "\n",
    "'''\n",
    "def get_bayes_ranking_test (df, threshold=0.95):\n",
    "  m=prep_Bayes(df)\n",
    "  required_columns = {'SAF', 'try_ratio', 'matched'}\n",
    "  if not required_columns.issubset(m.columns):\n",
    "    missing_cols = required_columns - set(m.columns)\n",
    "    raise ValueError(f\"Missing columns in DataFrame: {missing_cols}\")\n",
    "  m1=m[m['tag_count']>0]\n",
    "  X = m1[['SAF', 'try_ratio']].to_numpy()\n",
    "  y = m1['matched'].to_numpy()\n",
    "  scaler = MinMaxScaler()\n",
    "  X_scaled = scaler.fit_transform(X.reshape(-1, 1)).reshape(*X.shape)\n",
    "  X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.2, random_state=7)\n",
    "  gnb = GaussianNB()\n",
    "  gnb.fit(X_train, y_train)\n",
    "  y_pred = gnb.predict(X_test)\n",
    "  accuracy = accuracy_score(y_test, y_pred)\n",
    "  precision = precision_score(y_test, y_pred)\n",
    "  f1=f1_score(y_test, y_pred)\n",
    "  print(f\"✅ Gaussian Naive Bayes model ▶️ accuracy:{accuracy:.4f}, precision:{precision:.4f}, f1:{f1:.4f}\")\n",
    "  whole_pred=gnb.predict(X_scaled)\n",
    "  class_probabilities = gnb.predict_proba(X_scaled)\n",
    "  m2=m1.assign(pred=class_probabilities[:, 1])\n",
    "  m3=m2[m2['pred']>=threshold]\n",
    "  return m3\n",
    "\n",
    "\n",
    "#combine previous functions together to output a shortlisted .fasta\n",
    "\n",
    "def matching_ranking_to_fasta_v5 (mztab_path, fasta_df):\n",
    "  p=prep_mztab(mztab_path)\n",
    "  p1=casa_filter(p)\n",
    "  k=int(p1['nAA'].median())\n",
    "  m=matching_count_v5 (fasta_df, p1, k, chunk_size=10000)\n",
    "  m1=get_bayes_ranking_test (m)\n",
    "  seq_records = []\n",
    "  for index, row in m1.iterrows():\n",
    "    header_id = f\"{row['Description']}\"\n",
    "    sequence = Seq(row['Sequence'])\n",
    "    description = \"\"\n",
    "    seq_record = SeqRecord(sequence, id=header_id, description=description)\n",
    "    seq_records.append(seq_record)\n",
    "\n",
    "  output_fasta_filepath = mztab_path.replace('.mztab', '_matched.fasta')\n",
    "\n",
    "  with open(output_fasta_filepath, 'w') as output_file:\n",
    "    SeqIO.write(seq_records, output_file, 'fasta')\n",
    "  print(f\"🎊 Number of protein entries in the output fasta: {m1.shape[0]}\")\n",
    "\n",
    "\n",
    "\n",
    "#generate a de novo-first, experiment-specific .fasta for each input\n",
    "def process_all_mztab_files_v2 (folder_path, database_path):\n",
    "    mztab_filepaths = glob.glob(f\"{folder_path}/*.mztab\")\n",
    "    print(f\"🗂️ {len(mztab_filepaths)} file(s) collecting from {folder_path}...\")\n",
    "    fas=fasta_to_df(database_path)\n",
    "    fasta_df=pd.DataFrame.from_dict(fas)\n",
    "    print(f\"⬆️ {database_path} loaded\")\n",
    "    print(f\"📤 No. of proteins in the reference fasta: {fasta_df.shape[0]}\")\n",
    "\n",
    "    for mztab in mztab_filepaths:\n",
    "      print(f\"🚀 Processing file: {mztab}\")\n",
    "      matching_ranking_to_fasta_v5 (mztab, fasta_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "cDwhdpIFm1OW",
    "outputId": "6f901727-ae7f-4507-9667-c424cd49fdf6"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024-11-07 13:19:24.257880: I tensorflow/core/util/port.cc:153] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.\n",
      "2024-11-07 13:19:24.275841: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:485] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered\n",
      "2024-11-07 13:19:24.297059: E external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:8454] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered\n",
      "2024-11-07 13:19:24.303561: E external/local_xla/xla/stream_executor/cuda/cuda_blas.cc:1452] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered\n",
      "2024-11-07 13:19:24.319415: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.\n",
      "To enable the following instructions: AVX2 AVX512F AVX512_VNNI FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.\n",
      "2024-11-07 13:19:25.432563: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT\n",
      "Seed set to 454\n",
      "INFO: Downloading model weights file /root/.cache/casanovo/casanovo_v4_2_0_v4_2_0.ckpt from https://github.com/Noble-Lab/casanovo/releases/download/v4.2.0/casanovo_v4_2_0.ckpt\n",
      "100% 542M/542M [02:15<00:00, 4.19MB/s]\n",
      "INFO: Casanovo version 4.2.1\n",
      "INFO: Sequencing peptides from:\n",
      "INFO:   /content/drive/Shareddrives/MS_Palaeome/Odense_RAWs/23_Apr_2024/mzML/test/1007062_BLG_L_T0_1.mzML\n",
      "GPU available: True (cuda), used: True\n",
      "TPU available: False, using: 0 TPU cores\n",
      "HPU available: False, using: 0 HPUs\n",
      "INFO: Reading 1 files...\n",
      "/content/drive/Shareddrives/MS_Palaeome/Odense_RAWs/23_Apr_2024/mzML/test/1007062_BLG_L_T0_1.mzML: 100% 8648/8648 [00:06<00:00, 1307.40spectra/s]\n",
      "You are using a CUDA device ('NVIDIA A100-SXM4-40GB') that has Tensor Cores. To properly utilize them, you should set `torch.set_float32_matmul_precision('medium' | 'high')` which will trade-off precision for performance. For more details, read https://pytorch.org/docs/stable/generated/torch.set_float32_matmul_precision.html#torch.set_float32_matmul_precision\n",
      "LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]\n",
      "Predicting DataLoader 0: 100% 8/8 [02:04<00:00, 15.58s/it]\n",
      "INFO: DONE!\n",
      "2024-11-07 13:24:35.524909: I tensorflow/core/util/port.cc:153] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.\n",
      "2024-11-07 13:24:35.542444: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:485] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered\n",
      "2024-11-07 13:24:35.563645: E external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:8454] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered\n",
      "2024-11-07 13:24:35.570118: E external/local_xla/xla/stream_executor/cuda/cuda_blas.cc:1452] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered\n",
      "2024-11-07 13:24:35.585318: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.\n",
      "To enable the following instructions: AVX2 AVX512F AVX512_VNNI FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.\n",
      "2024-11-07 13:24:36.665905: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT\n",
      "Seed set to 454\n",
      "INFO: Model weights file /root/.cache/casanovo/casanovo_v4_2_0_v4_2_0.ckpt retrieved from local cache\n",
      "INFO: Casanovo version 4.2.1\n",
      "INFO: Sequencing peptides from:\n",
      "INFO:   /content/drive/Shareddrives/MS_Palaeome/Odense_RAWs/23_Apr_2024/mzML/test/1007063_BLG_L_T0_2.mzML\n",
      "GPU available: True (cuda), used: True\n",
      "TPU available: False, using: 0 TPU cores\n",
      "HPU available: False, using: 0 HPUs\n",
      "INFO: Reading 1 files...\n",
      "/content/drive/Shareddrives/MS_Palaeome/Odense_RAWs/23_Apr_2024/mzML/test/1007063_BLG_L_T0_2.mzML: 100% 8784/8784 [00:06<00:00, 1270.90spectra/s]\n",
      "You are using a CUDA device ('NVIDIA A100-SXM4-40GB') that has Tensor Cores. To properly utilize them, you should set `torch.set_float32_matmul_precision('medium' | 'high')` which will trade-off precision for performance. For more details, read https://pytorch.org/docs/stable/generated/torch.set_float32_matmul_precision.html#torch.set_float32_matmul_precision\n",
      "LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]\n",
      "Predicting DataLoader 0: 100% 8/8 [02:03<00:00, 15.43s/it]\n",
      "INFO: DONE!\n"
     ]
    }
   ],
   "source": [
    "#@title Run `Casanovo` for de novo\n",
    "\n",
    "if use_default:\n",
    "  folder = glob.glob(f\"{folder_path}/*.{file_type}\")\n",
    "  for instrument_file in folder:\n",
    "    output_path=instrument_file.replace(f\".{file_type}\", \".mztab\")\n",
    "    ! casanovo sequence {instrument_file} -v info -o {output_path}\n",
    "\n",
    "else:\n",
    "  folder = glob.glob(f\"{folder_path}/*.{file_type}\")\n",
    "  for instrument_file in folder:\n",
    "    output_path=instrument_file.replace(f\".{file_type}\", \".mztab\")\n",
    "    ! casanovo sequence {instrument_file} -m {model} -c {config} -v info -o {output_path}\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "CvDZIzfwrxbc",
    "outputId": "5114c954-2520-4fa8-e909-555fb99cf97f"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "uniprot_sprot.fasta.gz downloaded successfully.\n",
      "🗂️ 2 file(s) collecting from /content/drive/Shareddrives/MS_Palaeome/Odense_RAWs/23_Apr_2024/mzML/test...\n",
      "⬆️ uniprot_sprot.fasta loaded\n",
      "📤 No. of proteins in the reference fasta: 572214\n",
      "🚀 Processing file: /content/drive/Shareddrives/MS_Palaeome/Odense_RAWs/23_Apr_2024/mzML/test/1007062_BLG_L_T0_1.mztab\n",
      "📝 12809 tags regenerated. Starting matching...\n",
      "Completed! 2114 matched 👍 \n",
      "🧑‍💻 Start Bayes probabilistic ranking...\n",
      "✅ Gaussian Naive Bayes model ▶️ accuracy:0.9991, precision:0.4726, f1:0.6419\n",
      "🎊 Number of protein entries in the output fasta: 945\n",
      "🚀 Processing file: /content/drive/Shareddrives/MS_Palaeome/Odense_RAWs/23_Apr_2024/mzML/test/1007063_BLG_L_T0_2.mztab\n",
      "📝 13605 tags regenerated. Starting matching...\n",
      "Completed! 2022 matched 👍 \n",
      "🧑‍💻 Start Bayes probabilistic ranking...\n",
      "✅ Gaussian Naive Bayes model ▶️ accuracy:0.9989, precision:0.3700, f1:0.5382\n",
      "🎊 Number of protein entries in the output fasta: 944\n"
     ]
    }
   ],
   "source": [
    "#@title Convert `Casanovo` results to .fasta per experiment\n",
    "\n",
    "if use_SwissProt:\n",
    "  url = \"https://ftp.uniprot.org/pub/databases/uniprot/knowledgebase/complete/uniprot_sprot.fasta.gz\"\n",
    "  output_file = \"uniprot_sprot.fasta.gz\"\n",
    "  decompressed_file = \"uniprot_sprot.fasta\"\n",
    "  response = requests.get(url, stream=True)\n",
    "  if response.status_code == 200:\n",
    "    with open(output_file, 'wb') as f:\n",
    "      shutil.copyfileobj(response.raw, f)\n",
    "    print(f\"{output_file} downloaded successfully.\")\n",
    "  else:\n",
    "    print(f\"Failed to download {output_file}, status code: {response.status_code}\")\n",
    "  with gzip.open(output_file, 'rb') as f_in:\n",
    "    with open(decompressed_file, 'wb') as f_out:\n",
    "      shutil.copyfileobj(f_in, f_out)\n",
    "  sprot_path=\"uniprot_sprot.fasta\"\n",
    "  process_all_mztab_files_v2(folder_path, sprot_path)\n",
    "\n",
    "else:\n",
    "  process_all_mztab_files_v2(folder_path, database_path)\n"
   ]
  }
 ],
 "metadata": {
  "accelerator": "GPU",
  "colab": {
   "authorship_tag": "ABX9TyPviBaPleNLy6EHMo/BtU7R",
   "gpuType": "A100",
   "include_colab_link": true,
   "machine_shape": "hm",
   "mount_file_id": "1zYQy-qQw4qQuJJFziotDxoXyl8MHDf0r",
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3",
   "name": "python3"
  },
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
