{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "machine_shape": "hm",
      "gpuType": "V28",
      "mount_file_id": "15EyssAx15FkYLuUO0TueWN1f41hXmfyc",
      "authorship_tag": "ABX9TyOc2LZ6nS53sZDmiewEJxCL",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    },
    "accelerator": "TPU"
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/yc386/orthrus_metaproteomics/blob/main/orthrus_v110/orthrus_v110_pt2.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# **Orthrus pt2- [`Sage`](https://github.com/lazear/sage)**\n",
        "\n",
        "Please note: change to **TPU** runtime if RAM usage is expected to be high (due to files/PTMs/databases)"
      ],
      "metadata": {
        "id": "g529CpYsDdjc"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Add inputs for `SAGE` -> click `Runtime` -> `Run all`\n",
        "\n",
        "#@markdown **Parameters for `SAGE`**\n",
        "peak_folder = \"\" #@param {type:\"string\"}\n",
        "file_type=\"mzML\" #@param [\"mzML\", \"mgf\"]\n",
        "#@markdown - use the drop-down menu to choose the instrument file type\n",
        "json_file_path = '/content/drive/MyDrive/casanovo/sage/config_general_MQ_fixed_CAM_v1.json' #@param {type:\"string\"}\n",
        "#@markdown - a `Sage`-compatible `.json` file\n",
        "enzyme = \"KR\" #@param {type:\"string\"}\n",
        "\n",
        "#@markdown **`SAGE` PTM plus**\n",
        "#@markdown - Default `Sage` contains CAM (fixed) + variable mods: Oxidation(M), Deamidation(NQ)\n",
        "#@markdown - PTM plus up to 5 variable mods and CAM (cysteine carbamidomethylation) can be turned off\n",
        "use_PTM_plus = True #@param {type:\"boolean\"}\n",
        "static_CAM = True #@param {type:\"boolean\"}\n",
        "max_variable_mods= 3 #@param {type:\"number\"}\n",
        "missed_cleavages= 2 #@param {type:\"number\"}\n",
        "AA_1 = \"M\" #@param [\"None\", \"[\",\"]\",\"A\", \"C\", \"D\", \"E\", \"F\", \"G\", \"H\", \"I\", \"K\", \"L\", \"M\", \"N\", \"P\", \"Q\", \"R\", \"S\", \"T\", \"V\", \"W\", \"Y\"]\n",
        "AA_1_mod = 15.9949 #@param {type:\"number\"}\n",
        "AA_2 = \"P\" #@param [\"None\", \"[\",\"]\",\"A\", \"C\", \"D\", \"E\", \"F\", \"G\", \"H\", \"I\", \"K\", \"L\", \"M\", \"N\", \"P\", \"Q\", \"R\", \"S\", \"T\", \"V\", \"W\", \"Y\"]\n",
        "AA_2_mod = 15.9949 #@param {type:\"number\"}\n",
        "AA_3 = \"N\" #@param [\"None\", \"[\",\"]\",\"A\", \"C\", \"D\", \"E\", \"F\", \"G\", \"H\", \"I\", \"K\", \"L\", \"M\", \"N\", \"P\", \"Q\", \"R\", \"S\", \"T\", \"V\", \"W\", \"Y\"]\n",
        "AA_3_mod = 0.984016 #@param {type:\"number\"}\n",
        "AA_4 = \"Q\" #@param [\"None\", \"[\",\"]\",\"A\", \"C\", \"D\", \"E\", \"F\", \"G\", \"H\", \"I\", \"K\", \"L\", \"M\", \"N\", \"P\", \"Q\", \"R\", \"S\", \"T\", \"V\", \"W\", \"Y\"]\n",
        "AA_4_mod = 0.984016 #@param {type:\"number\"}\n",
        "AA_5 = \"None\" #@param [\"None\", \"[\",\"]\",\"A\", \"C\", \"D\", \"E\", \"F\", \"G\", \"H\", \"I\", \"K\", \"L\", \"M\", \"N\", \"P\", \"Q\", \"R\", \"S\", \"T\", \"V\", \"W\", \"Y\"]\n",
        "#@markdown - [ = n-terminal\n",
        "AA_5_mod = \t42.010565 #@param {type:\"number\"}\n"
      ],
      "metadata": {
        "id": "QSA5zC6q6knU",
        "cellView": "form"
      },
      "execution_count": 4,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title install dependencies\n",
        "%%time\n",
        "\n",
        "import os\n",
        "import shutil\n",
        "import glob\n",
        "import json\n",
        "\n",
        "if not os.path.isfile(\"Sage_READY\"):\n",
        "  print(\"installing conda...\")\n",
        "  os.system(\"wget -qnc https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh\")\n",
        "  os.system(\"bash ./Miniconda3-latest-Linux-x86_64.sh -b -f -p /usr/local\")\n",
        "  os.system(\"touch Sage_READY\")\n",
        "  os.system(f\"conda install -c bioconda -c conda-forge sage-proteomics -y -q\")\n",
        "\n",
        "\n",
        "'''\n",
        "Organise mztabs and instrument files in the same folder\n",
        "\n",
        "'''\n",
        "\n",
        "def organise_files (directory):\n",
        "    print(\"Organising files...\")\n",
        "\n",
        "    if not os.path.isdir(directory):\n",
        "        print(f\"The directory {directory} does not exist.\")\n",
        "        return\n",
        "\n",
        "\n",
        "\n",
        "    MS2_files = glob.glob(os.path.join(directory, f'*.{file_type}'))\n",
        "\n",
        "    for MS2 in MS2_files:\n",
        "\n",
        "        base_name = os.path.splitext(os.path.basename(MS2))[0]\n",
        "\n",
        "        new_folder_path = os.path.join(directory, base_name)\n",
        "\n",
        "        if not os.path.exists(new_folder_path):\n",
        "            os.makedirs(new_folder_path)\n",
        "            print(f\"Created folder: {new_folder_path}\")\n",
        "        else:\n",
        "            print(f\"Folder already exists: {new_folder_path}\")\n",
        "\n",
        "\n",
        "        MS2_path = os.path.join(new_folder_path, os.path.basename(MS2))\n",
        "        if not os.path.exists(MS2_path):\n",
        "            shutil.move(MS2, new_folder_path)\n",
        "            print(f\"Moved {MS2} to {new_folder_path}\")\n",
        "        else:\n",
        "            print(f\"MS2 file already exists in the destination: {MS2_path}\")\n",
        "\n",
        "\n",
        "        fasta_filename = f\"{base_name}_matched.fasta\"\n",
        "        fasta_file = os.path.join(directory, fasta_filename)\n",
        "\n",
        "\n",
        "        if os.path.exists(fasta_file):\n",
        "            new_fasta_path = os.path.join(new_folder_path, fasta_filename)\n",
        "            if not os.path.exists(new_fasta_path):\n",
        "                shutil.move(fasta_file, new_folder_path)\n",
        "                print(f\"Moved {fasta_file} to {new_folder_path}\")\n",
        "            else:\n",
        "                print(f\".fasta file already exists in the destination: {new_fasta_path}\")\n",
        "        else:\n",
        "            print(f\"No matching .fasta file found for {base_name}\")\n",
        "\n",
        "\n",
        "\"\"\"Generate and save a Sage configuration .json file.\"\"\"\n",
        "\n",
        "def get_sage_config(json_file_path, peak_folder, static_mods, new_mods,\n",
        "                    missed_cleavages, enzyme,\n",
        "                    min_len, max_len, max_variable_mods,\n",
        "                    output_config_path):\n",
        "\n",
        "    with open(json_file_path, 'r') as file:\n",
        "        json_data = json.load(file)\n",
        "\n",
        "        peak_files = glob.glob(peak_folder)\n",
        "        print(f\"üóÇÔ∏è {len(peak_files)} file(s) collected from {peak_folder}\")\n",
        "\n",
        "        json_data['mzml_paths'] = peak_files\n",
        "        json_data['database']['static_mods'] = static_mods\n",
        "        json_data['database']['variable_mods'] = new_mods\n",
        "        json_data['database']['enzyme']['missed_cleavages'] = missed_cleavages\n",
        "        json_data['database']['enzyme']['cleave_at']= enzyme\n",
        "        json_data['database']['enzyme']['min_len'] = min_len\n",
        "        json_data['database']['enzyme']['max_len'] = max_len\n",
        "        json_data['database']['max_variable_mods'] = max_variable_mods\n",
        "        json_data['database']['decoy_tag'] = \"rev_\"\n",
        "        json_data['database']['generate_decoys'] = True\n",
        "\n",
        "    with open(output_config_path, 'w') as f:\n",
        "        json.dump(json_data, f, indent=4)\n"
      ],
      "metadata": {
        "id": "VxKTRVsNs_Yv",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "cellView": "form",
        "outputId": "e9f486fd-7724-486a-a0cc-8c3a9de64aa0"
      },
      "execution_count": 2,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "installing conda...\n",
            "CPU times: user 143 ms, sys: 28.7 ms, total: 172 ms\n",
            "Wall time: 50.6 s\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Run Sage\n",
        "\n",
        "organise_files(peak_folder)\n",
        "folder_path = peak_folder\n",
        "\n",
        "if use_PTM_plus:\n",
        "    AAs = [AA_1, AA_2, AA_3, AA_4, AA_5]\n",
        "    mods = [AA_1_mod, AA_2_mod, AA_3_mod, AA_4_mod, AA_5_mod]\n",
        "    PTMs = {}\n",
        "\n",
        "    for AA, mod in zip(AAs, mods):\n",
        "        if AA != \"None\":\n",
        "            PTMs[AA] = [mod]\n",
        "\n",
        "    big_folder = glob.glob(f\"{folder_path}/*\")\n",
        "    for folder in big_folder:\n",
        "        if not os.path.isdir(folder):\n",
        "            continue\n",
        "\n",
        "        mzml_files = glob.glob(f\"{folder}/*.{file_type}\")\n",
        "        if not mzml_files:\n",
        "            continue\n",
        "\n",
        "        peak_path = mzml_files[0]\n",
        "        output_json = peak_path.replace(f\".{file_type}\", '.json')\n",
        "\n",
        "        json_file_path = json_file_path\n",
        "        missed_cleavages = missed_cleavages\n",
        "        enzyme = enzyme\n",
        "        min_len = 6\n",
        "        max_len = 30\n",
        "        max_variable_mods = max_variable_mods\n",
        "        static_mods = {\"C\": 57.021464} if static_CAM else {}\n",
        "\n",
        "        get_sage_config(\n",
        "            json_file_path, peak_path, static_mods, PTMs,\n",
        "            missed_cleavages, enzyme,\n",
        "            min_len, max_len,\n",
        "            max_variable_mods, output_json\n",
        "        )\n",
        "\n",
        "        fasta_files = glob.glob(f\"{folder}/*.fasta\")\n",
        "        if not fasta_files:\n",
        "            continue\n",
        "\n",
        "        fasta_path = fasta_files[0]\n",
        "        !sage {output_json} --fasta {fasta_path} \\\n",
        "            --write-pin --output_directory {folder}\n",
        "\n",
        "else:\n",
        "    big_folder = glob.glob(f\"{folder_path}/*\")\n",
        "    for folder in big_folder:\n",
        "        if not os.path.isdir(folder):\n",
        "            continue\n",
        "\n",
        "        mzml_files = glob.glob(f\"{folder}/*.{file_type}\")\n",
        "        if not mzml_files:\n",
        "            continue\n",
        "\n",
        "        peak_path = mzml_files[0]\n",
        "        output_json = peak_path.replace(f\".{file_type}\", '.json')\n",
        "        json_file_path = json_file_path\n",
        "        missed_cleavages = 2\n",
        "        enzyme = enzyme\n",
        "        min_len = 6\n",
        "        max_len = 30\n",
        "        max_variable_mods = 5\n",
        "        static_mods = {\"C\": 57.021464}\n",
        "        new_mods = {\"M\": [15.994915], \"N\": [0.984016], \"Q\": [0.984016]}\n",
        "\n",
        "        get_sage_config(\n",
        "            json_file_path, peak_path, static_mods, new_mods,\n",
        "            missed_cleavages, enzyme,\n",
        "            min_len, max_len,\n",
        "            max_variable_mods, output_json\n",
        "        )\n",
        "\n",
        "        fasta_files = glob.glob(f\"{folder}/*.fasta\")\n",
        "        if not fasta_files:\n",
        "            continue\n",
        "\n",
        "        fasta_path = fasta_files[0]\n",
        "        !sage {output_json} --fasta {fasta_path} \\\n",
        "            --write-pin --output_directory {folder}\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "cellView": "form",
        "id": "qDo9CbOsBWIU",
        "outputId": "9f6a8425-a4d0-443d-efed-a821eca0cdd3"
      },
      "execution_count": 5,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Organising files...\n",
            "The directory  does not exist.\n"
          ]
        }
      ]
    }
  ]
}